engine/access/rest/websockets/controller.go

```diff
@@ -39,62 +43,134 @@ func NewWebSocketController(
 func (c *Controller) HandleConnection(ctx context.Context) {
-	//TODO: configure the connection with ping-pong and deadlines
+
+	// configuring the connection with appropriate read/write deadlines and handlers.
+	err := c.configureKeepalive()
+	if err != nil {
+		// TODO: add error handling here
+		c.logger.Error().Err(err).Msg("error configuring keepalive connection")
+		c.shutdownConnection()
+		return
+	}
+
 	//TODO: spin up a response limit tracker routine
-	go c.readMessagesFromClient(ctx)
-	c.writeMessagesToClient(ctx)
+
+	// for track all goroutines and error handling
+	g, gCtx := errgroup.WithContext(ctx)
+
+	g.Go(func() error {
+		return c.readMessagesFromClient(gCtx)
+	})
+
+	g.Go(func() error {
+		return c.keepalive(gCtx)
+	})
+
+	g.Go(func() error {
+		return c.writeMessagesToClient(gCtx)
+	})
+
+	if err = g.Wait(); err != nil {
+		//TODO: add error handling here
+		c.logger.Error().Err(err).Msg("error detected in one of the goroutines")
+	}
+
+	c.shutdownConnection()
+}
+
+// configureKeepalive sets up the WebSocket connection with a read deadline
+// and a handler for receiving pong messages from the client.
+//
+// The function does the following:
+//  1. Sets an initial read deadline to ensure the server doesn't wait indefinitely
+//     for a pong message from the client. If no message is received within the
+//     specified `pongWait` duration, the connection will be closed.
+//  2. Establishes a Pong handler that resets the read deadline every time a pong
+//     message is received from the client, allowing the server to continue waiting
+//     for further pong messages within the new deadline.
+//
+// No errors are expected during normal operation.
+func (c *Controller) configureKeepalive() error {
+	// Set the initial read deadline for the first pong message
+	// The Pong handler itself only resets the read deadline after receiving a Pong.
+	// It doesn't set an initial deadline. The initial read deadline is crucial to prevent the server from waiting
+	// forever if the client doesn't send Pongs.
+	if err := c.conn.SetReadDeadline(time.Now().Add(PongWait)); err != nil {
+		return fmt.Errorf("failed to set the initial read deadline: %w", err)
+	}
+	// Establish a Pong handler which sets the handler for pong messages received from the peer.
+	c.conn.SetPongHandler(func(string) error {
+		return c.conn.SetReadDeadline(time.Now().Add(PongWait))
+	})
+
+	return nil
 }
 
 // writeMessagesToClient reads a messages from communication channel and passes them on to a client WebSocket connection.
 // The communication channel is filled by data providers. Besides, the response limit tracker is involved in
 // write message regulation
-func (c *Controller) writeMessagesToClient(ctx context.Context) {
-	//TODO: can it run forever? maybe we should cancel the ctx in the reader routine
+//
+// Expected errors during normal operation:
+// - context.Canceled if the client disconnected
```

Comment:
this doesn't return context errors anymore. this message indicates that it's safe to continue processing on error
```suggestion
// No errors are expected during normal operation. All errors are considered benign. 
```

Reply:
[cfad79f](https://github.com/onflow/flow-go/pull/6757/commits/cfad79fe97069d6acb12321bbd7666234c263b22)